Add merge arrow direction and update styling

kdumontnu opened the pull request:
Close https://github.com/go-gitea/gitea/issues/28522

~Adds some [negative margin](https://tailwindcss.com/docs/margin#using-negative-values) helper css classes using tailwind's [prefix syntax](https://tailwindcss.com/docs/configuration#prefix)~

### Before
![image](https://github.com/go-gitea/gitea/assets/12700993/5bdabf91-facd-41c8-8e36-e1535beb9409)

### After
![image](https://github.com/go-gitea/gitea/assets/12700993/a6f11f6f-9e64-45b6-b9d7-dfea53fbc6d7)

kdumontnu commented:
Unrelated CI failure

wxiaoguang commented on web_src/css/helpers.css:
```diff
@@ -158,6 +158,12 @@ Gitea's private styles use `g-` prefix.
 .gt-mt-4 { margin-top: 1rem !important; }
 .gt-mt-5 { margin-top: 2rem !important; }
 
+.-gt-mt-1 { margin-top: -.125rem !important; }
+.-gt-mt-2 { margin-top: -.25rem !important; }
+.-gt-mt-3 { margin-top: -.5rem !important; }
+.-gt-mt-4 { margin-top: -1rem !important; }
+.-gt-mt-5 { margin-top: -2rem !important; }
```
It doesn't seem to be tailwind style.

ps: I also prefer to use something else, instead of tricky negative margin -- if possible.

![image](https://github.com/go-gitea/gitea/assets/2114189/f92c284f-b5db-4b57-8d24-ce6bccc8e847)

kdumontnu replied:
@wxiaoguang My point was just that we seem to be following tailwind naming conventions, but with `gt-*` prefixed. In tailwind negative margin would be `-mt-4`. Are you suggesting `gt--mt-2` instead of `-gt-mt-2`?

I can use position instead of margin to remove a negative. I think it's less intuitive, but something like [this would work](https://play.tailwindcss.com/Vt1urArWoZ) if you like it better. I'd just need to generate the top/bottom position helper class definitions. 

BTW, why are we recreating tailwind? Is it because there are too many class conflicts?

wxiaoguang replied:
> So are you suggesting `gt--mt-2` instead of `-gt-mt-2`?

See the screenshot, I think it is `gt--mt-2` from the official example.

----

I haven't checked the details carefully and have no idea about the what's the "best" approach at the moment.

kdumontnu replied:
oh, you're saying we follow the `tw-*` [prefix syntax](https://tailwindcss.com/docs/configuration#prefix). I see... gross, but yeah I agree it's probably the best option given it more closely follows the tailing docs.

kdumontnu replied:
👍

jtran replied:
Tailwind's [prefix](https://tailwindcss.com/docs/configuration#prefix) link explicitly points out that it goes _before_ the prefix.

> The dash modifier for negative values should be added before your prefix, so -mt-8 would become -tw-mt-8 if you’ve configured tw- as your prefix:
>
> ```
> <div class="-tw-mt-8">
>   <!-- -->
> </div>
> ```

Also, in the [Tailwind Play screenshot](https://play.tailwindcss.com/zypvTUxU65) above, if you change `tw--bottom-4` to `-tw-bottom-4`, it behaves identically. Apparently it supports both.

kdumontnu replied:
Good point. I changed it back. This PR should be all set for review.

wxiaoguang commented:
My thoughts: https://github.com/go-gitea/gitea/pull/28523#discussion_r1431449455

> ps: I also prefer to use something else, instead of tricky negative margin -- if possible.

1. I do not think it's good to introduce these tricky negative margin helpers.
2. Why not just simply show the "left arrow"? Then the layout could also be simple and clear.

kdumontnu commented:
> My thoughts: [#28523 (comment)](https://github.com/go-gitea/gitea/pull/28523#discussion_r1431449455)
> 
> > ps: I also prefer to use something else, instead of tricky negative margin -- if possible.
> 
> 1. I do not think it's good to introduce these tricky negative margin helpers.

Negative margin is not inherently a problem. I need the two elements to overlap each other a bit. Otherwise, they look terrible.

The notation we’ve elected into is a bit tricky, but it is pretty well defined. (Imo the better strategy is just to use tailwind completely and not recreate it).

>    * At least, could it use the styles provided by the CSS class names?

I’m not sure what this means. I’m trying to follow the new helper class style. If you prefer, I can use the old style and just add a class to repo.css, but I thought we were migrating away from that.

> 2. Why not just simply show the "left arrow"? Then the layout could also be simple and clear.

I’m not sure what this means. Without the negative margin it looks terrible. With ascii left arrow it still looks bad. 

> (I won't block if most people would like to use the negative margin here)

kdumontnu commented:
If the negative margin definitions are weirding people out, I can use the old non-gt method for defining it, but what I've implemented is [what we've agreed to](https://github.com/go-gitea/gitea/blob/cdc33b29a012e61b42f192d79f9486fa8e21b2ed/docs/content/contributing/guidelines-frontend.en-us.md?plain=1#L50).

wxiaoguang commented:
> If the negative margin definitions are weirding people out, I can use the old non-gt method for defining it, but what I've implemented is [what we've agreed to](https://github.com/go-gitea/gitea/blob/cdc33b29a012e61b42f192d79f9486fa8e21b2ed/docs/content/contributing/guidelines-frontend.en-us.md?plain=1#L50).

I wrote these documents. But I would like to say "no" to these negative margins at the moment because I don't see they are "general" or "useful" helpers.

For the tailwind problem, maybe this issue could partially answer your questions: #23011

For example, I guess few people could understand this code:

```
<div class="code-comment-buttons gt-df gt-ac gt-fw gt-mt-3 gt-mb-2 gt-mx-3">
```

Why not just clearly define the styles for UI components/elements?

IMO these helpers are only for "fine-tuning" purpose: if some common/general/existing classes doesn't fit a special case, then use these helper functions to "fine tune".

----

I am not the frontend expert, nor a experienced tailwind library user. So as I said above: I won't block if most people would like to use the negative margin here.

kdumontnu commented:
> > If the negative margin definitions are weirding people out, I can use the old non-gt method for defining it, but what I've implemented is [what we've agreed to](https://github.com/go-gitea/gitea/blob/cdc33b29a012e61b42f192d79f9486fa8e21b2ed/docs/content/contributing/guidelines-frontend.en-us.md?plain=1#L50).
> 
> I wrote these documents. But I would like to say "no" (personally) to these negative margins at the moment because I don't see they are "general" or "useful" helpers.
> 
> For the tailwind problem, maybe this issue could partially answer your questions: #23011
> 
> For example, I guess few people could understand this code:
> 
> ```
> <div class="code-comment-buttons gt-df gt-ac gt-fw gt-mt-3 gt-mb-2 gt-mx-3">
> ```
> 
> Why not just clearly define the styles for UI components/elements?
> 
> IMO these helpers are only for "fine-tuning" purpose: if some common/general/existing classes don't fit a special case, then use these helper functions to "fine tune".

I'm not an expert either, but I believe what you've described above _is_ tailwind. So maybe that's not the direction we want to go? I'm just confused because I tried to follow the direction I believed we were moving the styling and I seem to have missed the boat. But that just seems to be where we're at with front-end right now.

> I am not the frontend expert, nor a experienced tailwind library user. So as I said above: I won't block if most people would like to use the negative margin helpers here.

> I wrote these documents.

I'm really glad you wrote & updated those docs (and your contributions in general). For better or worse though, it seems you are the owner of this document & strategy. Maybe we can work together to update it. Maybe gitea should put out a request for a front-end developer. Everyone who has looked at this has unfortunately said "I'm not the UI expert". It seems there is support for the feature - all I'm trying to do is overlap the two elements slightly so that they don't look terrible. Imo, that is why negative margin exists.

<img width="286" alt="image" src="https://github.com/go-gitea/gitea/assets/12700993/7edb6397-6680-436b-afb9-d0f0e2af6fbe">

I've changed the implementation here to use a general CSS class. We could do something similar with fomantic with `ui center aligned`, but ironically it would still require a definition for negative margin helper class.

wxiaoguang commented:
> There are two questions here I think I'm conflating:
>     1. Do we want to move to tailwind and use those definitions as the primary styling?
>     2. Do we want to avoid using negative margin (either in this implementation or general)?
> Once I have these answers I can build a solution.

For 2, I think they should be used if it is necessary, but maybe it is too early to introduce general helpers. There are already a lot of negative margins in code base, many of them are like "-1px" or some special values for hacking Fomantic UI margins.

For 1, I have no idea at the moment. If majority all agree to do so and have a clear&feasible plan and work together (to rewrite existing code & help new contributors to rewrite their PR), then it could go. Otherwise, I do not see why it would succeed.

lunny commented:
> There are two questions here I think I'm conflating:
> 
>     1. Do we want to move to tailwind and use those definitions as the primary styling?
> 
>     2. Do we want to avoid using negative margin (either in this implementation or general)?
> 
> 
> Once I have these answers I can build a solution.

I personally agree we should move to tailwind.
